{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "461c8150",
      "metadata": {},
      "outputs": [],
      "source": [
        "def readSequencesFromFastaFile(filepath):\n",
        "\n",
        "    reads = []\n",
        "    currentSeq = \"\"\n",
        "\n",
        "    f = open(filepath, \"r\")\n",
        "    for line in f:\n",
        "\n",
        "        if(line[0] == '>'):\n",
        "            if(currentSeq != \"\"):\n",
        "                reads.append(currentSeq)\n",
        "                currentSeq = \"\"\n",
        "        else:\n",
        "            currentSeq += line.strip(\"\\r\\n\")\n",
        "            \n",
        "    if(currentSeq != \"\"):\n",
        "        reads.append(currentSeq)\n",
        "\n",
        "    f.close()\n",
        "    return reads"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "1b0ffe5f",
      "metadata": {},
      "outputs": [],
      "source": [
        "def assemble(reads):\n",
        "    main_seq = reads.pop(0)\n",
        "    while(len(reads) > 1): \n",
        "        \n",
        "        (cover, r) = (0, -1)\n",
        "        \n",
        "        for i in range(1, len(reads)):\n",
        "\n",
        "            min_length = min(len(main_seq), len(reads[i]))\n",
        "            \n",
        "            # préfixe : on compare la début de la séquence principale avec la fin de la séquence reads[i] :\n",
        "            #               MAINSEQ -->\n",
        "            #           <-- READI \n",
        "            prefix_cover = -min_length + 1\n",
        "            while(prefix_cover < 0 and reads[i][len(reads[i]) + prefix_cover:] != main_seq[:abs(prefix_cover)]):\n",
        "                prefix_cover+=1\n",
        "\n",
        "            if(abs(prefix_cover) > abs(cover)): \n",
        "                (cover, r) = (prefix_cover, i)\n",
        "\n",
        "            # suffixe : on compare la fin de la séquence principale avec le début de la séquence reads[i] :\n",
        "            #           <-- MAINSEQ\n",
        "            #                 READI --> \n",
        "            sufix_cover = min_length - 1\n",
        "            while(sufix_cover > 0 and main_seq[len(main_seq) - sufix_cover:] != reads[i][:sufix_cover]):\n",
        "                sufix_cover-=1\n",
        "\n",
        "            if(sufix_cover > abs(cover)): \n",
        "                (cover, r) = (sufix_cover, i)\n",
        "            \n",
        "            # Méthode un peu sale pour sortir de la boucle for lorsqu'on trouve un chevauchement de la taille maximale \n",
        "            # (à savoir un chevauchement de la taille de la plus petite chaîne -1) :\n",
        "            if(abs(cover) == min_length - 1): break\n",
        "\n",
        "        if(cover != 0):\n",
        "            r_seq = reads.pop(r)\n",
        "            if(cover < 0):\n",
        "                main_seq = r_seq[:len(r_seq) - cover] + main_seq\n",
        "            else:\n",
        "                main_seq = main_seq[:len(main_seq) - cover] + r_seq\n",
        "        else:\n",
        "            return \"Erreur, au moins une séquence ne contient pas de chevauchement (de longueur > 0) dans les chaînes proposées\"\n",
        "    \n",
        "    return main_seq"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "8009da22",
      "metadata": {},
      "outputs": [],
      "source": [
        "def assembleOpti(reads):\n",
        "    main_seq = reads[0]\n",
        "    while(len(reads) > 1): \n",
        "        \n",
        "        max_overlap = 0\n",
        "        r = -1\n",
        "        reads2 = [main_seq]\n",
        "        \n",
        "        for i in range(1, len(reads)):\n",
        "\n",
        "            readi = reads[i]\n",
        "            min_length = min(len(main_seq), len(readi))\n",
        "\n",
        "            # On compare le début de la séquence principale avec la fin de la séquence reads[i] :\n",
        "            #              MAINSEQ --> \n",
        "            #          <-- READI\n",
        "            readi_slice = readi[len(readi) - min_length + 1:]\n",
        "            main_seq_slice = main_seq[:min_length - 1]\n",
        "\n",
        "            while(len(readi_slice) > 0 and readi_slice != main_seq_slice):\n",
        "                readi_slice = readi_slice[1:]\n",
        "                main_seq_slice = main_seq_slice[:-1]\n",
        "                \n",
        "\n",
        "            if(len(readi_slice) > abs(max_overlap)):\n",
        "                max_overlap = -len(readi_slice)\n",
        "                if(r != -1): \n",
        "                    reads2.append(reads[r])\n",
        "                r = i\n",
        "            \n",
        "            if(abs(max_overlap) == min_length - 1):\n",
        "                reads2.extend(reads[i+1:])\n",
        "                break\n",
        "\n",
        "            # On compare la fin de la séquence principale avec le début de la séquence reads[i] :\n",
        "            #           <-- MAINSEQ\n",
        "            #                 READI --> \n",
        "            readi_slice = readi[:min_length - 1]\n",
        "            main_seq_slice = main_seq[len(main_seq) - min_length + 1:]\n",
        "\n",
        "            while(len(readi_slice) > 0 and readi_slice != main_seq_slice):\n",
        "                readi_slice = readi_slice[:-1]\n",
        "                main_seq_slice = main_seq_slice[1:]\n",
        "\n",
        "            if(len(readi_slice) > abs(max_overlap)):\n",
        "                if(r != i and r!=-1): \n",
        "                    reads2.append(reads[r])\n",
        "                max_overlap = len(readi_slice)\n",
        "                r = i\n",
        "            \n",
        "            if(abs(max_overlap) == min_length - 1):\n",
        "                reads2.extend(reads[i+1:])\n",
        "                break\n",
        "\n",
        "            if(len(readi_slice) < abs(max_overlap) and r != i):\n",
        "                reads2.append(readi)\n",
        "\n",
        "        if(max_overlap != 0):\n",
        "            r_seq = reads[r]\n",
        "            if(max_overlap < 0):\n",
        "                main_seq = r_seq[:len(r_seq) + max_overlap] + main_seq\n",
        "            else:\n",
        "                main_seq = main_seq + r_seq[max_overlap:]\n",
        "        else:\n",
        "            # raise Exception(\"Erreur, au moins une séquence ne contient pas de chevauchement (de longueur > 0) dans les chaînes proposées\")\n",
        "            \n",
        "            print(\"Erreur : au moins une séquence ne contient pas de chevauchement (de longueur > 0) dans les chaînes proposées\")\n",
        "            return main_seq\n",
        "\n",
        "        reads = reads2\n",
        "    return main_seq"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "ea86e535",
      "metadata": {},
      "outputs": [],
      "source": [
        "# Test :\n",
        "\n",
        "# Lecture du fichier fasta des reads de E.Coli (small) :\n",
        "\n",
        "readsFasta = readSequencesFromFastaFile(\"./fasta/Escherichia_coli_SMALL_READS.fasta\")\n",
        "\n",
        "# print(readsFasta[0])\n",
        "# Alignement des séquences :\n",
        "result = assembleOpti(readsFasta)\n",
        "\n",
        "# Comparaison de la séquence obtenue avec la vraie séquence (donnée dans un des fichiers fasta)  \n",
        "# Lecture du fichier fasta du résultat d'alignement attendu E.Coli (small) :\n",
        "\n",
        "fastaResult = readSequencesFromFastaFile(\"./fasta/Escherichia_coli_SMALL.fasta\").pop()\n",
        "\n",
        "# Comparaison :\n",
        "\n",
        "print(\"Résultat obtenu  = \" + result)\n",
        "print(\"Résultat attendu = \" + fastaResult)\n",
        "print(\"\\n> Correspondance : \" + str(result == fastaResult))\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "c72a7c4f",
      "metadata": {},
      "outputs": [],
      "source": [
        "def assembleBruijn(reads, k):\n",
        "    nodes = {}\n",
        "\n",
        "    if k < round(len(reads[0])/2) or k> len(reads[0]) - 1: \n",
        "        k = len(reads[0]) - 1\n",
        "    \n",
        "    # utilisation d'un set pour éviter les redondances de kmers\n",
        "    kmers = set()\n",
        "    for r in reads:\n",
        "\n",
        "        while len(r) >= k:\n",
        "            kmers.add(r[:k])\n",
        "            r = r[1:]\n",
        "\n",
        "    # Add a kmer to the graph \n",
        "    # nodes are represented with 2 attributes:\n",
        "    # - next nodes (name)\n",
        "    # - previous nodes (name)\n",
        "    for kmer in kmers:\n",
        "        key1 = kmer[:k - 1]\n",
        "        key2 = kmer[1:]\n",
        "\n",
        "        #First node :\n",
        "        node1 = nodes.get(key1)\n",
        "        if not node1:\n",
        "            node1 = {'next': {}, 'previous': {}}\n",
        "        \n",
        "        if not node1['next'].get(key2) :\n",
        "            node1['next'][key2] = 0\n",
        "        node1['next'][key2] += 1\n",
        "\n",
        "        # Add second node :\n",
        "        node2 = nodes.get(key2)\n",
        "        if not node2:\n",
        "            node2 = {'next': {}, 'previous': {}}\n",
        "                  \n",
        "        if not node2['previous'].get(key1) :\n",
        "            node2['previous'][key1] = 0\n",
        "        node2['previous'][key1] += 1\n",
        "\n",
        "        nodes[key1] = node1\n",
        "        nodes[key2] = node2\n",
        "\n",
        "\n",
        "    # assemble the graph :\n",
        "    base_key = next(iter(nodes))\n",
        "    node = nodes[base_key]\n",
        "    sequence = base_key\n",
        "    \n",
        "    # run through nodes in both left and right directions\n",
        "    while len(node['previous']) > 0:\n",
        "        key = next(iter(node['previous']))\n",
        "        node = nodes.pop(key)\n",
        "        sequence = key[:1] + sequence\n",
        "\n",
        "    if len(node['previous']) > 1:\n",
        "        raise Exception(\"Erreur, au moins une séquence ne contient pas de chevauchement (de longueur > 0) dans les chaînes proposées\")\n",
        "\n",
        "    node = nodes.pop(base_key)\n",
        "    while len(node['next']) > 0:\n",
        "        key = next(iter(node['next']))\n",
        "        node = nodes.pop(key)\n",
        "        sequence += key[-1:]\n",
        "\n",
        "    if len(node['next']) > 1:\n",
        "        raise Exception(\"Erreur, au moins une séquence ne contient pas de chevauchement (de longueur > 0) dans les chaînes proposées\")\n",
        "    \n",
        "    if len(nodes) > 0:\n",
        "        raise Exception(\"Erreur, au moins une séquence ne contient pas de chevauchement (de longueur > 0) dans les chaînes proposées\")\n",
        "\n",
        "    return sequence\n",
        "k=100\n",
        "coroReads = readSequencesFromFastaFile('./fasta/MN908947.3_READS_MIXED.fasta')\n",
        "start_time = time.time()\n",
        "assembleBruijn(coroReads, k)\n",
        "print(\"--- k = \"+str(k)+\"   = \"+str(time.time() - start_time)+\" seconds ---\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "56f7a9e6",
      "metadata": {},
      "outputs": [],
      "source": [
        "# Comparaison du temps d'exécution en faisant varier k :\n",
        "# Tests temps de calcul :\n",
        "import time\n",
        "\n",
        "coroReads = readSequencesFromFastaFile('./fasta/MN908947.3_READS_MIXED.fasta')\n",
        " \n",
        "# k : 51 -> 100\n",
        "for k in range(51, 101):\n",
        "    start_time = time.time()\n",
        "    assembleBruijn(coroReads, k)\n",
        "    print(\"--- k = \"+str(k)+\"   = \"+str(time.time() - start_time)+\" seconds ---\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "cd21477f",
      "metadata": {},
      "outputs": [],
      "source": [
        "# Test :\n",
        "\n",
        "# Lecture du fichier fasta des reads de E.Coli (small) :\n",
        "readsFasta = readSequencesFromFastaFile(\"./fasta/Escherichia_coli_SMALL_READS.fasta\")\n",
        "# print(readsFasta[0])\n",
        "# Alignement des séquences :\n",
        "result = assembleBruijn(readsFasta, 0)\n",
        "\n",
        "# Comparaison de la séquence obtenue avec la vraie séquence (donnée dans un des fichiers fasta)  \n",
        "# Lecture du fichier fasta du résultat d'alignement attendu E.Coli (small) :\n",
        "fastaResult = readSequencesFromFastaFile(\"./fasta/Escherichia_coli_SMALL.fasta\").pop()\n",
        "# Comparaison :\n",
        "\n",
        "print(\"Résultat obtenu  = \" + result)\n",
        "print(\"Résultat attendu = \" + fastaResult)\n",
        "print(\"\\n> Correspondance : \" + str(result == fastaResult))"
      ]
    }
  ],
  "metadata": {
    "interpreter": {
      "hash": "755bca588c5ad61560da42c0cec298747f8b74f9b58b010e18f102b95c79ceda"
    },
    "kernelspec": {
      "display_name": "Python 3.10.4 64-bit",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.4"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}
